doc/languages-frameworks/python.section.md

```diff
@@ -336,21 +336,55 @@ Python versions or modules, which is why they don't go in `python3Packages`.
 
 A distinction is made between applications and libraries, however, sometimes a
 package is used as both. In this case the package is added as a library to
-`python-packages.nix` and as an application to `all-packages.nix`. To reduce
-duplication the `toPythonApplication` can be used to convert a library to an
-application.
+`python-packages.nix` and as an application to the top-level. To reduce
+duplication the `toPythonApplication` function can be used to convert a library
+to an application.
 
-The Nix expression shall use [`buildPythonPackage`](#buildpythonpackage-function) and be called from
-`python-packages.nix`. A reference shall be created from `all-packages.nix` to
-the attribute in `python-packages.nix`, and the `toPythonApplication` shall be
-applied to the reference:
+The Nix expression shall use [`buildPythonPackage`](#buildpythonpackage-function)
+and be called from `python-packages.nix`, as usual. A reference shall be created
+from the top-level to the attribute in `python-packages.nix`, and
+`toPythonApplication` shall be applied to that reference.
+
+At the current moment there are two methods for achieving this. Let's start with
+the legacy one.
+
+##### Legacy Method {#topythonapplication-function-legacy-method}
+
+The legacy method consists in registering a reference from `all-packages.nix`
+to the attribute in `python-packages.nix`, and applying `toPythonApplication` to
+that reference. Let's take `youtube-dl` as an example:
 
 ```nix
 {
   youtube-dl = with python3Packages; toPythonApplication youtube-dl;
 }
 ```
 
+::: {.note}
+The legacy method is deprecated and should be avoided for new packages; prefer
+the new method instead.
+:::
+
+##### New Method {#topythonapplication-function-new-method}
+
+The new method consists in registering a reference from `pkgs/by-name`
+hierarchy, in a `callPackage` fashion. Taking the above `youtube-dl` as an
+example, the reference is registered at
+`pkgs/by-name/yo/youtube-dl/package.nix`:
+
+```nix
+{ python3Packages }:
+
+with python3Packages; toPythonApplication youtube-dl
+```
+
```

Comment:
Document the way to ping the Python version of a Python application exposed this way for override interface consistency. (Feel free to rephrase it.)

````suggestion

In case the top-level Python application requires a specific interpreter version, specify the canonical `python3Packages` to use in `all-packages.nix`, instead of hard-coding the desired `python3*Packages`, to keep the [override](#sec-pkg-override) interface consistent.

:::{.example #ex-toPythonApplication-by-name-ping}
# Ping the interpreter version of a Python application exposed in `pkgs/by-name`.
```nix
{
  youtube-dl = callPackage ../by-name/yo/youtube-dl/package.nix {
    python3Packages = python312Packages;
  };
}
```
:::
````

Reply:
This is already documented in the [`pkgs/by-name` readme](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults). I thought about making the same suggestion, but IMO it's better to have that information deduplicated. Perhaps add a pointer to the readme instead?

Reply:
How about

```

In case the top-level Python application requires a specific interpreter version, [change the implicit default value](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults) of argument `python3Packages` to the desired package set, instead of hard-coding `python3*Packages` as an argument. This ensures consistent interpreter version [overriding](#sec-pkg-override) via `python3Package` argument.
```